File: src/routes/repositories.js
const express = require('express');
const router = express.Router();
const db = require('../models/db');
const { v4: uuidv4 } = require('uuid');

// GET repos for a project
router.get('/project/:projectId', async (req, res) => {
  try {
    const repos = await db.repositories.findAsync({ projectId: req.params.projectId });
    repos.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    for (const r of repos) {
      r.docCount = await db.documents.countAsync({ repositoryId: r._id });
    }

    res.json({ success: true, data: repos });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// GET single repo
router.get('/:id', async (req, res) => {
  try {
    const repo = await db.repositories.findOneAsync({ _id: req.params.id });
    if (!repo) return res.status(404).json({ success: false, error: 'Repository not found' });
    res.json({ success: true, data: repo });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// POST create repo
router.post('/', async (req, res) => {
  try {
    const { projectId, name, description, tags } = req.body;
    if (!projectId || !name?.trim()) return res.status(400).json({ success: false, error: 'projectId and name required' });

    const project = await db.projects.findOneAsync({ _id: projectId });
    if (!project) return res.status(404).json({ success: false, error: 'Project not found' });

    const repo = await db.repositories.insertAsync({
      _id: uuidv4(),
      projectId,
      name: name.trim(),
      description: description?.trim() || '',
      tags: tags || [],
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    });

    res.json({ success: true, data: repo });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// PUT update repo
router.put('/:id', async (req, res) => {
  try {
    const { name, description, tags } = req.body;
    await db.repositories.updateAsync(
      { _id: req.params.id },
      { $set: { name, description, tags, updatedAt: new Date().toISOString() } }
    );
    const updated = await db.repositories.findOneAsync({ _id: req.params.id });
    res.json({ success: true, data: updated });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

// DELETE repo + its docs
router.delete('/:id', async (req, res) => {
  try {
    await db.documents.removeAsync({ repositoryId: req.params.id }, { multi: true });
    await db.repositories.removeAsync({ _id: req.params.id }, {});
    res.json({ success: true });
  } catch (err) {
    res.status(500).json({ success: false, error: err.message });
  }
});

module.exports = router;

File: src/routes/repositories.py
"""
src/routes/repositories.py
CRUD endpoints for Repositories.
"""
import uuid
from datetime import datetime, timezone
from bottle import request, response
from src.models.database import get_conn


def now_iso():
    return datetime.now(timezone.utc).isoformat()


def register(app):

    @app.get('/api/repositories/project/<project_id>')
    def list_repos(project_id):
        with get_conn() as conn:
            rows = conn.execute("""
                SELECT r.id, r.project_id, r.name, r.description, r.tags,
                       r.created_at, r.updated_at,
                       COUNT(d.id) AS doc_count
                FROM repositories r
                LEFT JOIN documents d ON d.repository_id = r.id
                WHERE r.project_id = ?
                GROUP BY r.id
                ORDER BY r.created_at DESC
            """, (project_id,)).fetchall()
        return {'success': True, 'data': [dict(r) for r in rows]}

    @app.get('/api/repositories/<repo_id>')
    def get_repo(repo_id):
        with get_conn() as conn:
            row = conn.execute(
                'SELECT * FROM repositories WHERE id = ?', (repo_id,)
            ).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Repository not found'}
        return {'success': True, 'data': dict(row)}

    @app.post('/api/repositories')
    def create_repo():
        body = request.json or {}
        name = (body.get('name') or '').strip()
        project_id = body.get('project_id', '').strip()
        if not name or not project_id:
            response.status = 400
            return {'success': False, 'error': 'name and project_id are required'}
        repo_id = str(uuid.uuid4())
        ts = now_iso()
        tags = body.get('tags', '[]')
        if isinstance(tags, list):
            import json
            tags = json.dumps(tags)
        with get_conn() as conn:
            conn.execute(
                """INSERT INTO repositories (id, project_id, name, description, tags, created_at, updated_at)
                   VALUES (?, ?, ?, ?, ?, ?, ?)""",
                (repo_id, project_id, name,
                 body.get('description', ''), tags, ts, ts)
            )
            row = conn.execute('SELECT * FROM repositories WHERE id = ?', (repo_id,)).fetchone()
        return {'success': True, 'data': dict(row)}

    @app.put('/api/repositories/<repo_id>')
    def update_repo(repo_id):
        body = request.json or {}
        name = (body.get('name') or '').strip()
        if not name:
            response.status = 400
            return {'success': False, 'error': 'name is required'}
        ts = now_iso()
        tags = body.get('tags', '[]')
        if isinstance(tags, list):
            import json
            tags = json.dumps(tags)
        with get_conn() as conn:
            conn.execute(
                """UPDATE repositories
                   SET name=?, description=?, tags=?, updated_at=?
                   WHERE id=?""",
                (name, body.get('description', ''), tags, ts, repo_id)
            )
            row = conn.execute('SELECT * FROM repositories WHERE id = ?', (repo_id,)).fetchone()
        if not row:
            response.status = 404
            return {'success': False, 'error': 'Repository not found'}
        return {'success': True, 'data': dict(row)}

    @app.delete('/api/repositories/<repo_id>')
    def delete_repo(repo_id):
        with get_conn() as conn:
            conn.execute('DELETE FROM repositories WHERE id = ?', (repo_id,))
        return {'success': True}
